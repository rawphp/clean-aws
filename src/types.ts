import { EventEmitter } from 'events';

export interface IResourceCleaner extends EventEmitter {
  list: () => Promise<object>;
  remove: (list: IRegionResources) => Promise<number>;
  toString: () => string;
}

export interface ICleanOptions {
  profile: string;
  region: string;
  resourceFile: string;
  dryRun: boolean;
  [key: string]: any;
}

// document types

export type ValidRegion = 'ap-southeast-2' | 'eu-west-1';

export interface IMasterResourceList {
  'us-east-1'?: IRegionResources;
  'us-east-2'?: IRegionResources;
  'us-west-1'?: IRegionResources;
  'us-west-2'?: IRegionResources;

  'ap-south-1'?: IRegionResources;

  'ap-northeast-1'?: IRegionResources;
  'ap-northeast-2'?: IRegionResources;
  'ap-northeast-3'?: IRegionResources;
  'ap-southeast-1'?: IRegionResources;
  'ap-southeast-2'?: IRegionResources;
  'ca-central-1'?: IRegionResources;
  'cn-north-1'?: IRegionResources;
  'cn-northwest-1'?: IRegionResources;

  'eu-central-1'?: IRegionResources;
  'eu-west-1'?: IRegionResources;
  'eu-west-2'?: IRegionResources;
  'eu-west-3'?: IRegionResources;

  'sa-east-1'?: IRegionResources;
  'us-gov-east-1'?: IRegionResources;
  'us-gov-west-1'?: IRegionResources;
}

export type IRegionResources =
  & ISQSList
  & ISNSList
  & IBucketsList
  & IIAMList
  & IEC2List
  & IDynamoDbTableList
  & IDataPipelineList
  & ICloudWatchList
  & ICloudFormationList;

export interface ISQSList {
  region: string;
  profile: string;
  queues: string[];
}

export interface ISNSList {
  region: string;
  profile: string;
  topics: string[];
  subscriptions: string[];
}

export interface IBucketsList {
  region: string;
  profile: string;
  buckets: string[];
}

export interface IIAMList {
  region: string;
  profile: string;
  policies: string[];
  roles: string[];
  accessKeys: string[];
  users: string[];
}

export interface IEC2List {
  region: string;
  profile: string;
  ec2s: string[];
}

export interface IDynamoDbTableList {
  region: string;
  profile: string;
  tables: string[];
}

export interface IDataPipelineList {
  region: string;
  profile: string;
  dataPipelines: string[];
}

export interface ICloudWatchList {
  region: string;
  profile: string;
  logGroups: string[];
}

export interface ICloudFormationList {
  region: string;
  profile: string;
  cloudFormationStacks: string[];
}
