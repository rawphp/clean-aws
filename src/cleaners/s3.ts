import * as AWS from 'aws-sdk';
import * as BPromise from 'bluebird';
import * as colors from 'colors/safe';
import { IResourceCleaner, IS3Options } from '../types';
import { Provider } from './Provider';

export class S3 extends Provider implements IResourceCleaner {
  protected s3: AWS.S3;
  protected buckets: string[];

  public constructor(options: IS3Options) {
    super('S3', options);

    this.buckets = options.buckets || [];
    this.s3 = options.s3 || new AWS.S3({
      region: options && options.region ? options.region : undefined,
    });
  }

  public getData(): any {
    return this.buckets;
  }

  public async list(): Promise<boolean> {
    console.log('[S3] Listing Buckets');

    this.emit('listStarted', { resource: this, region: this.region });

    try {
      const response = await this.s3.listBuckets().promise();

      if (!response || !response.Buckets) {
        return false;
      }

      this.buckets = response && response.Buckets ? response.Buckets.map((bucket: AWS.S3.Bucket) => bucket.Name as string) : [];

      this.emit('listCompleted', { resource: this, region: this.region, data: { buckets: this.buckets } });

      return true;
    } catch (error) {
      console.error(colors.red(error.message));

      return false;
    }
  }

  public async remove(): Promise<boolean> {
    console.log('[S3] Deleting Buckets');

    try {
      this.emit('deleteStarted', { resource: this, region: this.region, data: { buckets: this.buckets } });

      if (this.dryRun) {
        console.log('[S3] Dry Run');

        return true;
      }

      if (!this.buckets) {
        throw new Error('Bucket list has not been set');
      }

      await BPromise.map(this.buckets, async (bucket: string) => {
        try {
          let token: string | undefined;
          let versionMarker: string | undefined;
          let keyMarker: string | undefined;

          this.emit('deletionStarted', { region: this.region, name: bucket });

          do {
            const listResponse: AWS.S3.ListObjectsV2Output = await this.s3.listObjectsV2({
              Bucket: bucket,
              ContinuationToken: token,
            }).promise();

            // delete objects
            await BPromise.map(listResponse.Contents || [], async (obj) => {
              try {
                await this.s3.deleteObject({
                  Bucket: bucket,
                  Key: obj.Key,
                } as AWS.S3.DeleteObjectRequest).promise();
                process.stdout.write('.');
              } catch (error) {
                console.error(error.message);
              }
            });

            token = listResponse.NextContinuationToken;
          } while (token);

          do {
            const listVersionsResponse = await this.s3.listObjectVersions({
              Bucket: bucket,
              KeyMarker: keyMarker,
              VersionIdMarker: versionMarker,
            }).promise();

            // delete versions
            await BPromise.map(listVersionsResponse.Versions || [], async (obj) => {
              try {
                await this.s3.deleteObject({
                  Bucket: bucket,
                  Key: obj.Key,
                  VersionId: obj.VersionId,
                } as AWS.S3.DeleteObjectRequest).promise();
                process.stdout.write('.');
              } catch (error) {
                console.error(error.message);
              }
            }, { concurrency: 1000 });

            // delete delete markers
            await BPromise.map(listVersionsResponse.DeleteMarkers || [], async (obj) => {
              try {
                await this.s3.deleteObject({
                  Bucket: bucket,
                  Key: obj.Key,
                  VersionId: obj.VersionId,
                } as AWS.S3.DeleteObjectRequest).promise();
                process.stdout.write('.');
              } catch (error) {
                console.error(error.message);
              }
            }, { concurrency: 1000 });

            versionMarker = listVersionsResponse.NextVersionIdMarker;
            keyMarker = listVersionsResponse.KeyMarker;
          } while (versionMarker);

          try {
            // delete bucket
            await this.s3.deleteBucket({
              Bucket: bucket,
            }).promise();
            process.stdout.write('.');
          } catch (error) {
            console.error(colors.red('Delete bucket error'), bucket, error.message);
          }

          this.emit('deletionCompleted', { region: this.region, name: bucket });
        } catch (error) {
          console.error(colors.red(error.message));
        }
      }, { concurrency: 5 });

      this.emit('deleteCompleted', { resource: this, region: this.region, data: this.buckets });

      return true;
    } catch (error) {
      if (error.message === 'The specified bucket does not exist') {
        return true;
      }

      console.error(colors.red(error.message));

      return false;
    }
  }
}
