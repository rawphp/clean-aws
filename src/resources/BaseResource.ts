import { EventEmitter } from 'events';
import { ICleanOptions } from '../types';

export abstract class BaseResource extends EventEmitter {
  public region: string;
  public profile: string;
  public dryRun: boolean;
  protected name: string;

  public constructor(name: string, options: ICleanOptions) {
    super();

    this.name = name;
    this.region = options.region;
    this.profile = options.profile;
    this.dryRun = options.dryRun;
  }

  public toString() {
    return `${this.name}-${this.profile}-${this.region}`;
  }
}
