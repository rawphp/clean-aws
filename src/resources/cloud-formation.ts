import * as AWS from 'aws-sdk';
import { isBoolean } from 'util';
import { ICleanOptions, ICloudFormationList, IResourceCleaner } from '../types';
import { BaseResource } from './BaseResource';

export class CloudFormation extends BaseResource implements IResourceCleaner {
  protected cloudFormation: AWS.CloudFormation;

  public constructor(options: ICleanOptions) {
    super('CloudFormation', options);

    this.cloudFormation =
      options.cloudFormation ||
      new AWS.CloudFormation({
        region: options && options.region ? options.region : '',
      });
  }

  public async list(): Promise<object> {
    console.log('[CloudFormation] Listing Stacks', this.region);

    this.emit('listStarted', { resource: this, region: this.region });

    try {
      let next: boolean | string | undefined = true;

      const file: ICloudFormationList = {
        region: this.region,
        profile: this.profile,
        cloudFormationStacks: [],
      };

      while (next) {
        if (isBoolean(next)) {
          next = undefined;
        }

        const stacks = await this.cloudFormation
          .describeStacks({
            NextToken: next as string,
          })
          .promise();

        if (stacks && stacks.Stacks) {
          stacks.Stacks.forEach((stack: AWS.CloudFormation.Stack) => {
            file.cloudFormationStacks.push(stack.StackName);
          });
        }

        next = false;

        if (stacks.NextToken) {
          console.log('has NextToken', stacks.NextToken);

          next = stacks.NextToken;
        }
      }

      return file;
    } catch (error) {
      console.error(error);

      throw error;
    } finally {
      this.emit('listCompleted', { resource: this, region: this.region });
    }
  }

  public async remove({ cloudFormationStacks }: { cloudFormationStacks: string[] }): Promise<number> {
    try {
      const processes = cloudFormationStacks.map((stack: string) => {
        return this.cloudFormation
          .deleteStack({
            StackName: stack,
          })
          .promise();
      });

      await Promise.all(processes);

      return 0;
    } catch (error) {
      console.error(error);

      return 1;
    }
  }
}
