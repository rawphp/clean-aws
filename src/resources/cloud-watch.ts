import * as AWS from 'aws-sdk';
import { ICleanOptions, ICloudWatchList, IResourceCleaner } from '../types';
import { BaseResource } from './BaseResource';

export class CloudWatch extends BaseResource implements IResourceCleaner {
  protected cloudWatch: AWS.CloudWatchLogs;

  public constructor(options: ICleanOptions) {
    super('CloudWatch', options);

    this.cloudWatch =
      options.cloudWatch ||
      new AWS.CloudWatchLogs({
        region: options && options.region ? options.region : undefined,
      });
  }

  public async list(): Promise<object> {
    console.log('[CloudWatch] Listing LogGroups');

    this.emit('listStarted', { resource: this, region: this.region });

    try {
      const response = await this.cloudWatch.describeLogGroups().promise();

      if (!response || !response.logGroups) {
        return { region: this.region, logGroups: [] };
      }

      const file: ICloudWatchList = {
        region: this.region,
        profile: this.profile,
        logGroups:
          response && response.logGroups
            ? response.logGroups.map((group: AWS.CloudWatchLogs.LogGroup) => group.logGroupName as string)
            : [],
      };

      return file;
    } catch (error) {
      console.error(error);

      throw error;
    } finally {
      this.emit('listCompleted', { resource: this, region: this.region });
    }
  }

  public async remove({ logGroups }: { logGroups: string[] }): Promise<number> {
    try {
      const processes = logGroups.map((logGroup: string) => {
        return this.cloudWatch
          .deleteLogGroup({
            logGroupName: logGroup,
          })
          .promise();
      });

      await Promise.all(processes);

      return 0;
    } catch (error) {
      console.error(error);

      return 1;
    }
  }
}
