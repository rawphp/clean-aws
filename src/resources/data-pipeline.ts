import * as AWS from 'aws-sdk';
import { ICleanOptions, IDataPipelineList, IResourceCleaner } from '../types';
import { BaseResource } from './BaseResource';

const availability = ['us-west-2', 'eu-west-1', 'ap-southeast-2', 'ap-northeast-1'];

export class DataPipeline extends BaseResource implements IResourceCleaner {
  protected dataPipeline: AWS.DataPipeline;

  public constructor(options: ICleanOptions) {
    super('DataPipeline', options);

    this.dataPipeline =
      options.dataPipeline ||
      new AWS.DataPipeline({
        region: options && options.region ? options.region : '',
      });
  }

  public async list(): Promise<object> {
    console.log('[DataPipeline] Listing Stacks', this.region);

    this.emit('listStarted', { resource: this, region: this.region });

    try {
      const file: IDataPipelineList = {
        region: this.region,
        profile: this.profile,
        dataPipelines: [],
      };

      if (!availability.includes(this.region)) {
        return file;
      }

      const pipelines = await this.dataPipeline.listPipelines().promise();

      if (pipelines && pipelines.pipelineIdList) {
        pipelines.pipelineIdList.forEach((pipeline: AWS.DataPipeline.PipelineIdName) => {
          file.dataPipelines.push(pipeline.id as string);
        });
      }

      return file;
    } catch (error) {
      console.error(error);

      throw error;
    } finally {
      this.emit('listCompleted', { resource: this, region: this.region });
    }
  }

  public async remove({ dataPipelines }: { dataPipelines: string[] }): Promise<number> {
    try {
      const processes = dataPipelines.map((pipelineId: string) => {
        return this.dataPipeline
          .deletePipeline({
            pipelineId,
          })
          .promise();
      });

      await Promise.all(processes);

      return 0;
    } catch (error) {
      console.error(error);

      return 1;
    }
  }
}
