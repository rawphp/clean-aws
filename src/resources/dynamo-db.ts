import * as AWS from 'aws-sdk';
import { ICleanOptions, IDynamoDbTableList, IResourceCleaner } from '../types';
import { BaseResource } from './BaseResource';

export class DynamoDb extends BaseResource implements IResourceCleaner {
  protected dynamoDb: AWS.DynamoDB;

  public constructor(options: ICleanOptions) {
    super('DynamoDb', options);

    this.dynamoDb =
      options.dynamoDb ||
      new AWS.DynamoDB({
        region: options && options.region ? options.region : undefined,
      });
  }

  public async list(): Promise<object> {
    console.log('[DynamoDb] Listing DynamoDb Tables');

    this.emit('listStarted', { resource: this, region: this.region });

    try {
      const response = await this.dynamoDb.listTables().promise();

      if (!response || !response.TableNames) {
        return { buckets: [] };
      }

      const file: IDynamoDbTableList = {
        region: this.region,
        profile: this.profile,
        tables:
          response && response.TableNames
            ? response.TableNames.map((table: AWS.DynamoDB.TableName) => table as string)
            : [],
      };

      return file;
    } catch (error) {
      console.error(error);

      throw error;
    } finally {
      this.emit('listCompleted', { resource: this, region: this.region });
    }
  }

  public async remove({ tables }: { tables: string[] }): Promise<number> {
    try {
      const processes = tables.map((name: string) => {
        return this.dynamoDb
          .deleteTable({
            TableName: name,
          })
          .promise();
      });

      await Promise.all(processes);

      return 0;
    } catch (error) {
      console.error(error);

      return 1;
    }
  }
}
