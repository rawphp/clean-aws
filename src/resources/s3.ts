import * as AWS from 'aws-sdk';
import { IBucketsList, ICleanOptions, IResourceCleaner } from '../types';
import { BaseResource } from './BaseResource';

export class S3 extends BaseResource implements IResourceCleaner {
  protected s3: AWS.S3;

  public constructor(options: ICleanOptions) {
    super('S3', options);

    this.s3 =
      options.s3 ||
      new AWS.S3({
        region: options && options.region ? options.region : undefined,
      });
  }

  public async list(): Promise<object> {
    console.log('[S3] Listing Buckets');

    this.emit('listStarted', { resource: this, region: this.region });

    try {
      const response = await this.s3.listBuckets().promise();

      if (!response || !response.Buckets) {
        return { buckets: [] };
      }

      const file: IBucketsList = {
        region: this.region,
        profile: this.profile,
        buckets:
          response && response.Buckets ? response.Buckets.map((bucket: AWS.S3.Bucket) => bucket.Name as string) : [],
      };

      return file;
    } catch (error) {
      console.error(error);

      throw error;
    } finally {
      this.emit('listCompleted', { resource: this, region: this.region });
    }
  }

  public async remove({ buckets }: { buckets: string[] }): Promise<number> {
    try {
      if (this.dryRun) {
        console.log('Dry Run');
        console.log('Deleting', buckets);

        return 0;
      }

      const processes = buckets.map((bucket: string) => {
        return this.s3
          .deleteBucket({
            Bucket: bucket,
          })
          .promise();
      });

      await Promise.all(processes);

      return 0;
    } catch (error) {
      console.error(error);

      return 1;
    }
  }
}
