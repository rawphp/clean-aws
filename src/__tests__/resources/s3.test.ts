import * as AWS from 'aws-sdk-mock';
import { S3 } from '../../resources/s3';
import { IResourceCleaner } from '../../types';
import * as listBucketsJson from '../fixtures/list-buckets.json';

describe('S3 resources', () => {
  let resource: IResourceCleaner;

  const options = {
    profile: 'default',
    region: 'ap-southeast-2',
    dryRun: false,
    resourceFile: 'resources.json',
  };

  beforeEach(() => {
    AWS.mock('S3', 'listBuckets', listBucketsJson);

    resource = new S3(options);
  });

  afterEach(() => {
    AWS.restore('S3');
  });

  test('list buckets', async () => {
    const buckets = await resource.list();

    expect(buckets).toMatchSnapshot();
  });
});
