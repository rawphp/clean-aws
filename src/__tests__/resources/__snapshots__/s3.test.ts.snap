// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`S3 resources list buckets 1`] = `
Object {
  "buckets": Array [
    "cleaner-integration-stub-serverlessdeploymentbuck-y58aik0p2irv",
    "collection-point-service-payloads-abc-ap-southeast-2",
    "collection-point-service-serverlessdeploymentbuck-jdu7llhd88nm",
    "data-crawler-abc-inventory-ap-southeast-2",
    "data-crawler-abc-serverlessdeploymentbucket-1chqsvxeus55v",
    "data-crawler-abc-test-backup-ap-southeast-2",
    "data-service-abc-ap-southeast-2-datawarehouse",
    "data-service-abc-serverlessdeploymentbucket-zhgxjvm38fpo",
    "dpd-uk-abc-serverlessdeploymentbucket-tijgx0o9pe46",
    "elasticbeanstalk-ap-southeast-2-156949620760",
    "elasticbeanstalk-eu-west-1-156949620760",
    "elasticsearch-tdev-serverlessdeploymentbucket-1eqritkg6slib",
    "elasticsearch-tdev-serverlessdeploymentbucket-gyw87icosqoy",
    "elasticsearch-tdev-serverlessdeploymentbucket-xpa6xx7p0or9",
    "elasticsearch-abc-serverlessdeploymentbucket-1s1am5x4a7zca",
    "fedex-v1-abc-serverlessdeploymentbucket-1jo1b7akuqnz2",
    "magento-shipping-service-serverlessdeploymentbuck-1fte8ja73ruq0",
    "providence-auth-abc-serverlessdeploymentbucket-1sivfa9m3eyqu",
    "rate-templates-ap-southeast-2-abc",
    "report-service-abc-ap-southeast-2-storage",
    "report-service-abc-serverlessdeploymentbucket-29h8lzsfl5md",
    "schemas-service-abc-ap-southeast-2",
    "schemas-service-abc-serverlessdeploymentbucket-1lsjeq6tcuj8o",
    "storage-v2-tdev-backup-ap-southeast-2",
    "storage-v2-tdev-backup-eu-west-1",
    "storage-v2-tdev-backup-us-east-2",
    "storage-v2-tdev-lfs-ap-southeast-2",
    "storage-v2-tdev-lfs-eu-west-1",
    "storage-v2-tdev-lfs-us-east-2",
    "storage-v2-abc-backup-ap-southeast-2",
    "storage-v2-abc-lfs-ap-southeast-2",
    "abc-backup",
    "abc-magento-portal",
    "temando-account-service-serverlessdeploymentbuck-1dknczv5nxq25",
    "temando-account-service-serverlessdeploymentbuck-kmjk3akcwaoj",
    "temando-billing-service-serverlessdeploymentbuck-1i5oguho0rvt5",
    "temando-cleaner-service-abc-ap-southeast-2-storage",
    "temando-console-review-cypress-te-6qz0p7-data",
    "temando-console-review-cypress-te-6qz0p7-web",
    "temando-console-review-pipeline-f-ujoi1e-data",
    "temando-console-review-pipeline-f-ujoi1e-web",
    "temando-console-review-revised-cy-5n10mu-data",
    "temando-console-review-revised-cy-5n10mu-web",
    "temando-console-review-abc-cypres-5f9zil-data",
    "temando-console-review-abc-cypres-5f9zil-web",
    "temando-console-review-test-cypre-i6mcit-data",
    "temando-console-review-test-cypre-i6mcit-web",
    "temando-console-abc-data",
    "temando-console-abc-web",
    "temando-document-service-abc-ap-southeast-2-bucket",
    "temando-documentation-storage-abc-ap-southeast-2",
    "temando-nginx-proxy-abc-cf",
    "temando-service-registry-serverlessdeploymentbuck-1q80fhjl9cvil",
    "temando-service-registry-serverlessdeploymentbuck-u8wqdykdgo4n",
    "temando-service-registry-tdev-state",
    "temando-service-registry-abc-state",
    "temando-session-service-serverlessdeploymentbuck-1ki7u1vw3o3y",
    "temando-sls-deployments-tdev-ap-southeast-2",
    "temando-sls-deployments-tdev-eu-west-1",
    "temando-sls-deployments-tdev-us-east-2",
    "temando-sls-deployments-abc-ap-southeast-2",
    "temando-sls-deployments-abc-eu-west-1",
    "temando-stream-service-t-serverlessdeploymentbuck-1jwbedisqj69t",
    "temando-swagger-registry-serverlessdeploymentbuck-zyt5s16dor73",
    "temando-swagger-registry-abc-storage",
    "temando-tracking-service-abc-store",
    "ups-v3-abc-serverlessdeploymentbucket-f7h6ngs32no",
    "vpc-infrastructure-abc-serverlessdeploymentbucket-1qa0tel4hc2rg",
  ],
  "profile": "default",
  "region": "ap-southeast-2",
}
`;
